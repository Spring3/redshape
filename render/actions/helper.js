const axios = require('electron').remote.require('../common/request');

function IssueFilter () {
  const filter = {};

  this.issue = (ids) => {
    filter.issue_id = Array.isArray(ids) ? ids.join(',') : ids;
    return this;
  };

  this.project = (id) => {
    filter.project_id = id;
    return this;
  };

  this.subproject = (id) => {
    filter.subprojectId = id;
    return this;
  };

  this.tracker = (id) => {
    filter.tracker_id = id;
    return this;
  };

  this.status = ({ open, closed }) => {
    if (open && closed) {
      filter.status_id = '*';
    } else {
      filter.status_id = open ? 'open' : 'closed';
    }
    return this;
  };

  this.assignee = (id) => {
    filter.assigned_to_id = id;
    return this;
  };

  this.author = (id) => {
    filter.author_id = id;
    return this;
  };

  this.title = (str) => {
    if (str) {
      filter.subject = encodeURIComponent(`~${str.trim()}`);
    }
    return this;
  };

  this.due = (date) => {
    filter.due_date = date.toISOString();
    return this;
  };

  this.createdAt = (date) => {
    filter.created_on = date.toISOString();
    return this;
  };

  this.createdBetween = (startDate, endDate) => {
    filter.created_on = encodeURIComponent(`><${startDate.toISOString()}|${endDate.toISOString()}`);
    return this;
  };

  this.priority = (id) => {
    filter.priority_id = id;
    return this;
  };

  this.sort = (by, direction) => {
    if (by && direction) {
      const column = by.indexOf('.') !== -1
        ? by.substring(0, by.indexOf('.'))
        : by;
      filter.sort = `${column}:${direction}`;
    }
    return this;
  };

  this.build = () => ({ ...filter });
}

const request = ({
  url,
  method = 'GET',
  data,
  headers,
  query,
  id
}) => {
  const requestConfig = {
    url,
    method,
    headers,
    id
  };

  if (!['GET', 'DELETE'].includes(method)) {
    requestConfig.data = typeof data === 'string' ? data : JSON.stringify(data);
  }

  if (query) {
    requestConfig.params = query;
  }

  return axios.authorizedRequest(requestConfig);
};

const login = ({
  redmineEndpoint,
  url,
  headers
}) => axios.request({
  baseURL: redmineEndpoint,
  timeout: 20000,
  headers: headers || {},
  url,
  method: 'GET'
}).then((res) => {
  const { api_key } = res.data.user || {};
  if (api_key) {
    axios.initialize(redmineEndpoint, api_key);
  }
  return { data: res.data };
});

const logout = () => axios.reset();

const notify = {
  start: (type, info = {}) => ({ type, status: 'START', info }),
  ok: (type, data, info = {}) => ({ type, data, status: 'OK', info }),
  nok: (type, data, info = {}) => ({ type, data, status: 'NOK', info }),
  cancel: (type, info = {}) => ({ type, status: 'CANCELLED', info })
};

export {
  IssueFilter,
  notify,
  login,
  logout
};

export default request;
