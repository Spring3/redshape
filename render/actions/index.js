import issueActions from './issue.actions';
import timeEntryActions from './timeEntry.actions';
import settingsActions from './settings.actions';

export default {
  issue: issueActions,
  timeEntry: timeEntryActions,
  settings: settingsActions
};
