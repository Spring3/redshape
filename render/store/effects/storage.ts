import { ipcRenderer, remote } from 'electron';
import type { Context } from 'overmind';
import { Response, StorageAction } from '../../../types';

const crypto = remote.require('crypto');

type SaveArgs = {
  settings: Context['state']['settings'];
  currentUser: Context['state']['users']['currentUser'];
};

const saveActiveSession = ({ settings, currentUser }: SaveArgs): Promise<Response> => {
  const id = crypto.randomBytes(10).toString('hex');
  const { endpoint, ...appSettings } = settings;

  return new Promise((resolve) => {
    ipcRenderer.send('storage', JSON.stringify({
      action: StorageAction.SAVE,
      payload: {
        user: {
          id: currentUser?.id,
          firstName: currentUser?.firstName,
          lastName: currentUser?.lastName,
          createdOn: currentUser?.createdOn
        },
        endpoint,
        settings: {
          ...appSettings
        }
      },
      id
    }));

    ipcRenderer.once(`storage:${id}`, (event, response) => {
      console.log('storage:save', response);
      resolve(response);
    });
  });
};

type GetSessionArgs = {
  token: string;
  endpoint: string;
};

const getSession = (payload: GetSessionArgs): Promise<Response<Context['state']['settings'] | undefined>> => {
  const id = crypto.randomBytes(10).toString('hex');

  return new Promise((resolve) => {
    ipcRenderer.send('storage', JSON.stringify({ action: StorageAction.READ, payload, id }));

    ipcRenderer.once(`storage:${id}`, (event, response) => {
      console.log('storage:read', response);
      const settings = response.success
        ? {
          endpoint: response.payload.redmine.endpoint,
          ...response.payload.settings
        }
        : undefined;

      resolve({
        success: response.success,
        error: response.error,
        payload: settings
      });
    });
  });
};

const resetActiveSession = (): Promise<Response> => {
  const id = crypto.randomBytes(10).toString('hex');

  return new Promise((resolve) => {
    ipcRenderer.send('storage', JSON.stringify({ action: StorageAction.RESET, id }));

    ipcRenderer.once(`storage:${id}`, (event, response) => {
      console.log('storage:reset', response);
      resolve(response);
    });
  });
};

export {
  saveActiveSession,
  getSession,
  resetActiveSession
};
