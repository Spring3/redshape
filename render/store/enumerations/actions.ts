import type { IAction, Context } from 'overmind';
import { Activity, Response } from '../../../types';

const fetchActivities: IAction<void, Promise<Response<Activity[]>>> = async ({ effects, state }: Context) => {
  const response = await effects.mainProcess.request({
    payload: {
      method: 'GET',
      route: 'enumerations/time_entry_activities.json',
    }
  });

  if (response.success) {
    state.enumerations.activities = response.data.activities as Activity[] ?? [];
    return {
      success: true,
      data: state.enumerations.activities,
    };
  }

  return {
    success: false,
    error: response.error
  };
};

export {
  fetchActivities
};
