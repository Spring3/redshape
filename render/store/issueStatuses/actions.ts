import { IAction, Context } from 'overmind';
import { IssueStatus, PaginatedActionResponse } from '../../../types';
import { indexById } from '../helpers';

const getAll: IAction<void, Promise<PaginatedActionResponse<IssueStatus>>> = async ({ effects, state }: Context) => {
  const response = await effects.mainProcess.request({
    payload: {
      method: 'GET',
      route: 'issue_statuses.json'
    }
  });

  if (response.success) {
    state.issueStatuses.byId = {
      ...state.issueStatuses,
      ...indexById(response.payload.issueStatuses)
    };

    return {
      success: true,
      data: {
        items: response.payload.issueStatuses,
        total: response.payload.total,
        limit: response.payload.limit,
        offset: response.payload.offset
      },
      hasMore: response.payload.total > (response.payload.offset + response.payload.issueStatuses.length)
    };
  }

  return {
    success: false,
    data: {
      items: [],
      total: 0,
      limit: response.payload.limit,
      offset: response.payload.offset
    },
    hasMore: false,
    error: response.error
  };
};

export {
  getAll
};
