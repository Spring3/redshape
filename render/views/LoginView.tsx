import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import { Formik } from 'formik';
import Joi from '@hapi/joi';
import { useHistory } from 'react-router-dom';

import { Input, Label } from '../components/Input';
import Button from '../components/Button';
import ErrorMessage from '../components/ErrorMessage';
import Copyrights from '../components/Copyrights';
import DragArea from '../components/DragArea';
import { useOvermindActions, useOvermindState } from '../store';

const Container = styled.div`
  display: grid;
  grid-template-rows: repeat(4, 25vh);
  grid-template-columns: repeat(6, minmax(100px, 1fr));
  align-items: center;
  justify-items: center;
`;

const LoginForm = styled.form`
  padding: 40px;
  grid-column: 2 / 6;
  grid-row: 2 / 4;
  min-width: 300px;
  min-height: 500px;
`;

const Headline = styled.h1`
  text-align: center;
  font-size: 40px;
  color: ${props => props.theme.main};
`;

const CopyrightsContainer = styled.div`
  grid-row: 4;
  grid-column: 2 / 6;
  align-self: end;
  margin-bottom: 20px;
`;

const SubmitButton = styled(Button)`
  padding: 10px 5px;
  margin: 25px auto 0px auto;
`;

type ValidateArgs = {
  apiKey: string;
  username: string;
  password: string;
  redmineEndpoint: string;
}

const LoginView = () => {
  const [useApiKey, setUseApiKey] = useState(false);
  const [loginError, setLoginError] = useState<Error>();

  const actions = useOvermindActions();
  const state = useOvermindState();

  const history = useHistory();

  useEffect(() => {
    if (state.users.currentUser?.id) {
      history.push('/app/summary');
    }
  }, [state.users.currentUser]);

  const validate = ({
    apiKey, username, password, redmineEndpoint
  }: ValidateArgs) => {
    const errors = {
      redmineEndpoint: Joi.string()
        .uri()
        .required()
        .validate(redmineEndpoint)
    };
    if (useApiKey) {
      // @ts-expect-error needs proper typing
      errors.apiKey = Joi.string()
        .required()
        .validate(apiKey);
    } else {
      // @ts-expect-error needs proper typing
      errors.username = Joi.string()
        .required()
        .validate(username);
      // @ts-expect-error needs proper typing
      errors.password = Joi.string()
        .required()
        .validate(password);
    }
    const results = {};
    for (const [prop, validation] of Object.entries(errors)) {
      if (validation.error) {
        // @ts-expect-error needs proper typing
        results[prop] = validation.error.message.replace('value', prop);
      }
    }
    return results;
  };

  const onSubmit = (values: any, { setSubmitting }: { setSubmitting: (b: boolean) => void }) => {
    actions.users.login({ ...values }).then(({ success, error }) => {
      setSubmitting(false);
      if (error) {
        setLoginError(error);
      } else if (success) {
        history.push('/app/summary');
      }
    });
  };

  const onToggleLoginMode = () => {
    setUseApiKey(use => !use);
  };

  return (
    <Container>
      <DragArea />
      <Formik
        initialValues={{
          username: '',
          password: '',
          apiKey: '',
          redmineEndpoint: ''
        }}
        validate={validate}
        onSubmit={onSubmit}
      >
        {({
          values, errors, touched, handleChange, handleBlur, handleSubmit, isSubmitting
        }) => (
          <LoginForm onSubmit={handleSubmit}>
            <Headline>Redshape</Headline>

            <Label label="Redmine Endpoint" htmlFor="redmineEndpoint">
              <Input
                name="redmineEndpoint"
                placeholder="https://redmine.example.com"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.redmineEndpoint}
              />
            </Label>
            <ErrorMessage show={Boolean(errors.redmineEndpoint && touched.redmineEndpoint)}>
              {errors.redmineEndpoint}
            </ErrorMessage>

            <Label label="Login mode" htmlFor="loginMode">
              <label>
                <Input
                  type="checkbox"
                  name="loginMode"
                  checked={useApiKey}
                  onChange={onToggleLoginMode}
                />
                <span>Use API Key</span>
              </label>
            </Label>
            {useApiKey ? (
              <>
                <Label label="API Key" htmlFor="apiKey">
                  <Input
                    type="text"
                    name="apiKey"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.apiKey}
                  />
                </Label>
                <ErrorMessage show={Boolean(errors.apiKey && touched.apiKey)}>{errors.apiKey}</ErrorMessage>
              </>
            ) : (
              <>
                <Label label="Login" htmlFor="username">
                  <Input
                    type="text"
                    name="username"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.username}
                  />
                </Label>
                <ErrorMessage show={Boolean(errors.username && touched.username)}>
                  {errors.username}
                </ErrorMessage>
                <Label label="Password" htmlFor="password">
                  <Input
                    type="password"
                    name="password"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.password}
                  />
                </Label>
                <ErrorMessage show={Boolean(errors.password && touched.password)}>
                  {errors.password}
                </ErrorMessage>
              </>
            )}

            <SubmitButton type="submit" disabled={isSubmitting} block={true}>
              Submit
            </SubmitButton>
            <ErrorMessage show={!!loginError}>{loginError && loginError.message}</ErrorMessage>
          </LoginForm>
        )}
      </Formik>
      <CopyrightsContainer>
        <Copyrights />
      </CopyrightsContainer>
    </Container>
  );
};

export default LoginView;
