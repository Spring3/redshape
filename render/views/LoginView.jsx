import React, { Component } from 'react';
import _ from 'lodash';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled, { withTheme } from 'styled-components';
import { Formik } from 'formik';
import Joi from 'joi';
import { withRouter } from 'react-router-dom';
import GithubCircleIcon from 'mdi-react/GithubCircleIcon';

import actions from '../actions';

import { Input, Label } from '../components/Input';
import Button from '../components/Button';
import ErrorMessage from '../components/ErrorMessage';
import Link from '../components/Link';
import Copyrights from '../components/Copyrights';
import DragArea from '../components/DragArea';

const Container = styled.div`
  display: grid;
  grid-template-rows: repeat(4, 25vh);
  grid-template-columns: repeat(6, minmax(100px, 1fr));
  align-items: center;
  justify-items: center;
`;

const LoginForm = styled.form`
  padding: 40px;
  grid-column: 2 / 6;
  grid-row: 2 / 4;
  min-width: 300px;
`;

const Headline = styled.h1`
  text-align: center;
  font-size: 40px;
  color: ${props => props.theme.main};
`;

const CopyrightsContainer = styled.div`
  grid-row: 4;
  grid-column: 2 / 6;
  align-self: end;
  margin-bottom: 20px;
`;

const SubmitButton = styled(Button)`
  padding: 10px 5px;
  margin: 25px auto 0px auto;
`;

class LoginView extends Component {  
  componentDidMount() {
    const { userId, api_key } = this.props;
    if (userId && api_key) {
      this.props.history.push('/app/summary');
    }
  }

  validate = ({ username, password, redmineEndpoint }) => {
    const errors = {
      username: Joi.validate(username, Joi.string().required()),
      password: Joi.validate(password, Joi.string().required()),
      redmineEndpoint: Joi.validate(redmineEndpoint, Joi.string().uri().required())
    };
    const results = {};
    for (const [prop, validation] of Object.entries(errors)) {
      if (validation.error) {
        results[prop] = validation.error.message.replace('value', prop);
      }
    }
    return results;
  };

  onSubmit = (values, { setSubmitting }) => {
    const { checkLogin } = this.props;
    checkLogin(values).then(() => {
      const { loginError, userId } = this.props;
      if (!loginError && userId) {
        this.props.history.push('/app/summary');
      }
      setSubmitting(false);
    });
  }

  render() {
    const { loginError } = this.props; 
    return (
      <Container>
        <DragArea />
        <Formik
          initialValues={{ username: '', password: '', redmineEndpoint: '' }}
          validate={this.validate}
          onSubmit={this.onSubmit}
        >
          {({
            values,
            errors,
            touched,
            handleChange,
            handleBlur,
            handleSubmit,
            isSubmitting
          }) => (
            <LoginForm onSubmit={handleSubmit}>
              <Headline>Redshape</Headline>
              <Label
                label="Login"
                htmlFor="username"
              >
                <Input
                  type="text"
                  name="username"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.username}
                />
              </Label>
              <ErrorMessage show={errors.username && touched.username}>
                {errors.username}
              </ErrorMessage>
              <Label
                label="Password"
                htmlFor="password"
              >
                <Input
                  type="password"
                  name="password"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.password}
                />
              </Label>
              <ErrorMessage show={errors.password && touched.password}>
                {errors.password}
              </ErrorMessage>
              <Label
                label="Redmine Endpoint"
                htmlFor="redmineEndpoint"
              >
                <Input
                  name="redmineEndpoint"
                  placeholder="https://redmine.example.com"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.redmineEndpoint}
                />
              </Label>
              <ErrorMessage show={errors.redmineEndpoint && touched.redmineEndpoint}>
                {errors.redmineEndpoint}
              </ErrorMessage>
              <SubmitButton
                type="submit"
                disabled={isSubmitting}
                block={true}
              >
                Submit
              </SubmitButton>
              <ErrorMessage show={!!loginError}>
                {loginError && loginError.message}
              </ErrorMessage>
            </LoginForm>
          )}
        </Formik>
        <CopyrightsContainer>
          <Copyrights />
        </CopyrightsContainer>
      </Container>
    );
  }
}

LoginView.propTypes = {
  userId: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  api_key: PropTypes.string,
  loginError: PropTypes.instanceOf(Error),
  checkLogin: PropTypes.func.isRequired
}

const mapStateToProps = state => ({
  userId: state.user.id,
  api_key: state.user.api_key,
  loginError: state.user.loginError
});

const mapDispatchToProps = dispatch => ({
  checkLogin: credentials => dispatch(actions.user.checkLogin(credentials))
});

export default withTheme(connect(mapStateToProps, mapDispatchToProps)(withRouter(LoginView)));
