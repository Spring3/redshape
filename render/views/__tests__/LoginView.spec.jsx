import React from 'react';
import { HashRouter, Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { render, cleanup, wait } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';
import MockAdapter from 'axios-mock-adapter';

import actions from '../../actions';
import { USER_LOGIN } from '../../actions/user.actions';
import { notify } from '../../actions/helper';
import storage from '../../../common/storage';
import axios from '../../../common/request';

import theme from '../../theme';
import LoginView from '../LoginView';

let axiosMock;

const mockStore = configureStore([thunk]);

describe('Login view', () => {
  beforeAll(() => {
    axiosMock = new MockAdapter(axios.default);
  });

  afterEach(() => {
    cleanup();
    axiosMock.reset();
    storage.clear();
  });

  afterAll(() => {
    axiosMock.restore();
    axios.reset();
  });

  it('should match the snapshot', () => {
    const store = mockStore({ user: {} });
    const tree = renderer.create(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <LoginView />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match the snapshot when using api key', () => {
    const store = mockStore({ user: {} });
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <LoginView />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    );

    let inputs = wrapper.find('input');
    const checkbox = inputs.at(1);
    checkbox.simulate('change', { target: { checked: true } });
    inputs = wrapper.find('input');
    expect(inputs.length).toBe(3);
    const apiKeyInput = wrapper.find('input[name="apiKey"]');
    expect(apiKeyInput).toBeTruthy();

    expect(wrapper).toMatchSnapshot();
  });

  it('should validate the username', () => {
    const store = mockStore({ user: {} });
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <LoginView />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    );

    const usernameInput = wrapper.find('input[name="username"]');
    expect(usernameInput).toBeTruthy();
    usernameInput.simulate('focus');
    usernameInput.simulate('blur');
    let error = wrapper.findWhere((item) => item.text() === '"username" is not allowed to be empty');
    expect(error).toBeTruthy();
    usernameInput.simulate('change', { target: { value: 'username' }, preventDefault: () => {} });
    usernameInput.simulate('blur');
    error = wrapper.findWhere((item) => item.text() === '"username" is not allowed to be empty');
    expect(error.exists()).toBeFalsy();
    expect(store.getActions().length).toBe(0);
  });

  it('should validate the password', () => {
    const store = mockStore({ user: {} });
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <LoginView />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    );
    const passwordInput = wrapper.find('input[type="password"]');
    expect(passwordInput).toBeTruthy();
    passwordInput.simulate('focus');
    passwordInput.simulate('blur');
    let error = wrapper.findWhere((item) => item.text() === '"password" is not allowed to be empty');
    expect(error).toBeTruthy();
    passwordInput.simulate('change', { target: { value: 'password' } });
    error = wrapper.findWhere((item) => item.text() === '"password" is not allowed to be empty');
    expect(error.exists()).toBeFalsy();
    expect(store.getActions().length).toBe(0);
  });

  it('should validate the redmine domain input', () => {
    const store = mockStore({ user: {} });
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <LoginView />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    );
    const redmineInput = wrapper.find('input[name="redmineEndpoint"]');
    expect(redmineInput).toBeTruthy();
    redmineInput.simulate('focus');
    redmineInput.simulate('blur');
    let error = wrapper.findWhere((item) => item.text() === '"redmineEndpoint" is not allowed to be empty');
    expect(error).toBeTruthy();
    redmineInput.simulate('change', { target: { value: 'url' } });
    error = wrapper.findWhere((item) => item.text() === '"redmineEndpoint" must be a valid uri');
    expect(error).toBeTruthy();
    redmineInput.simulate('change', { target: { value: 'https://redmine.domain' } });
    expect(error.exists()).toBeFalsy();
    expect(store.getActions().length).toBe(0);
  });

  it('should redirect to the AppView if user credentials already exist', async () => {
    const storeData = {
      user: {
        id: 1,
        firstname: 'firstname',
        lastname: 'lastname',
        api_key: 'api_key',
        redmineEndpoint: 'https://redmine.domain'
      }
    };
    const store = mockStore(storeData);
    const historyMock = {
      location: {
        pathname: '/'
      },
      listen: () => () => {},
      push: jest.fn()
    };

    render(
      <Provider store={store}>
        <Router history={historyMock}>
          <ThemeProvider theme={theme}>
            <LoginView />
          </ThemeProvider>
        </Router>
      </Provider>
    );
    await wait(() => {
      expect(store.getActions().length).toBe(0);
      store.clearActions();
    });
  });
});

describe('[integration] LoginView', () => {
  beforeAll(() => {
    axiosMock = new MockAdapter(axios.default);
  });

  afterEach(() => {
    cleanup();
    axiosMock.reset();
    storage.clear();
  });

  afterAll(() => {
    axiosMock.restore();
    axios.reset();
  });

  it('should not make a redmine api request if the form has errors', (done) => {
    const store = mockStore({ user: {} });
    const storageSetSpy = jest.spyOn(storage, 'set');
    axiosMock.onGet('/users/current.json').reply(() => Promise.resolve([200]));
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <LoginView />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    );
    const form = wrapper.find('LoginView');
    expect(form.exists()).toBeTruthy();
    const submitButton = wrapper.find('button[type="submit"]');
    expect(submitButton.exists()).toBeTruthy();
    submitButton.simulate('submit');
    setTimeout(() => {
      expect(wrapper.findWhere((item) => /is not allowed to be empty/.test(item.text())).exists()).toBe(true);
      expect(storageSetSpy).not.toHaveBeenCalled();
      expect(store.getActions().length).toBe(0);
      storageSetSpy.mockRestore();
      done();
    }, 100);
  });

  it('should make a redmine api request on submit', (done) => {
    const store = mockStore({ user: {} });
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <LoginView />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    );
    const form = wrapper.find('LoginView');
    expect(form.exists()).toBeTruthy();

    const userData = {
      id: '123',
      firstname: 'test',
      lastname: 'user',
      api_key: 'api_key'
    };

    expect(wrapper.exists('button[type="submit"]')).toBe(true);

    const inputs = wrapper.find('input');
    expect(inputs.length).toBe(4);
    const redmineEndpointInput = inputs.at(0);
    const checkbox = inputs.at(1);
    const usernameInput = inputs.at(2);
    const passwordInput = inputs.at(3);
    expect(checkbox.prop('name')).toBe('loginMode');
    expect(usernameInput.prop('name')).toBe('username');
    expect(passwordInput.prop('name')).toBe('password');
    expect(redmineEndpointInput.prop('name')).toBe('redmineEndpoint');

    const returnedValues = {
      username: 'username',
      password: 'password',
      redmineEndpoint: 'https://redmine.domain'
    };

    axiosMock.onGet('/users/current.json').reply(() => Promise.resolve([200, { user: userData }]));

    wrapper.find('input[name="username"]').simulate('change', { persist: () => {}, target: { value: returnedValues.username, name: 'username' } });
    wrapper.find('input[name="password"]').simulate('change', { persist: () => {}, target: { value: returnedValues.password, name: 'password' } });
    wrapper.find('input[name="redmineEndpoint"]').simulate('change', { persist: () => {}, target: { name: 'redmineEndpoint', value: returnedValues.redmineEndpoint } });
    expect(wrapper.find('input[name="username"]').prop('value')).toBe(returnedValues.username);
    expect(wrapper.find('input[name="password"]').prop('value')).toBe(returnedValues.password);
    expect(wrapper.find('input[name="redmineEndpoint"]').prop('value')).toBe(returnedValues.redmineEndpoint);
    wrapper.find('LoginView__LoginForm').simulate('submit', { preventDefault: () => {} });

    setTimeout(() => {
      wrapper.update();
      expect(wrapper.findWhere((item) => /is not allowed to be empty/.test(item.text())).length).toBe(0);

      expect(axiosMock.history.get.length).toBe(1);
      expect(axiosMock.history.get[0].url).toBe(`${returnedValues.redmineEndpoint}/users/current.json`);
      expect(axiosMock.history.get[0].headers.Authorization).toBe(`Basic ${btoa(`${returnedValues.username}:${returnedValues.password}`)}`);

      expect(store.getActions().length).toBe(3);
      expect(store.getActions()[0]).toEqual(notify.start(USER_LOGIN));
      expect(store.getActions()[1]).toEqual(notify.ok(USER_LOGIN, {
        user: {
          ...userData,
          redmineEndpoint: returnedValues.redmineEndpoint
        }
      }));

      store.clearActions();
      done();
    }, 200);
  });

  it('should display the error if one raised during the request', async () => {
    // simulating error since reducers are not called
    const store = mockStore({ user: { loginError: new Error('Something went wrong') } });
    const loginActionSpy = jest.spyOn(actions.user, 'checkLogin');
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <LoginView />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    );
    const form = wrapper.find('LoginView');
    expect(form.exists()).toBeTruthy();

    const expectedError = new Error('Something went wrong');

    const returnedValues = {
      username: 'username',
      password: 'password',
      redmineEndpoint: 'https://redmine.domain'
    };

    axiosMock.onGet(`${returnedValues}/users/current.json`).reply(() => Promise.reject(expectedError));

    wrapper.find('input[name="username"]').simulate('change', { persist: () => {}, target: { value: returnedValues.username, name: 'username' } });
    wrapper.find('input[name="password"]').simulate('change', { persist: () => {}, target: { value: returnedValues.password, name: 'password' } });
    wrapper.find('input[name="redmineEndpoint"]').simulate('change', { persist: () => {}, target: { name: 'redmineEndpoint', value: returnedValues.redmineEndpoint } });
    expect(wrapper.find('input[name="username"]').prop('value')).toBe(returnedValues.username);
    expect(wrapper.find('input[name="password"]').prop('value')).toBe(returnedValues.password);
    expect(wrapper.find('input[name="redmineEndpoint"]').prop('value')).toBe(returnedValues.redmineEndpoint);

    wrapper.find('button[type="submit"]').simulate('click', { preventDefault: () => {} });

    setTimeout(() => {
      wrapper.update();
      expect(loginActionSpy).toHaveBeenCalledWith(returnedValues);
      expect(axiosMock.history.get.length).toBe(1);
      expect(axiosMock.history.get[0].url).toBe(`${returnedValues.redmineEndpoint}/users/current.json`);
      expect(axiosMock.history.get[0].headers.Authorization).toBe(`Basic ${btoa(`${returnedValues.username}:${returnedValues.password}`)}`);

      expect(wrapper.findWhere((item) => /Something went wrong/.test(item.text())).length).toBeGreaterThan(0);
      const reduxActions = store.getActions();
      expect(reduxActions.length).toBe(2);
      expect(reduxActions[0]).toEqual(notify.start(USER_LOGIN));
      expect(reduxActions[1]).toEqual(notify.nok(USER_LOGIN, expectedError));

      store.clearActions();
      loginActionSpy.mockRestore();
    }, 200);
  });
});
