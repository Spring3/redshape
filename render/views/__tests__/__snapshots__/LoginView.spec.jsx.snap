// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login view should match the snapshot 1`] = `
.c6 {
  display: block;
  width: 100%;
  border-radius: 3px;
  padding: 5px 10px;
  box-sizing: border-box;
  font-size: 14px;
  min-height: 35px;
  outline: none;
  font-weight: bold;
  -webkit-transition: background .2s;
  transition: background .2s;
  border: 1px solid #A0A0A0;
  color: #FF7079;
  background: white;
}

.c6:hover {
  border: 1px solid #FF7079;
}

.c6:focus {
  border-color: #FF7079;
  box-shadow: 0px 0px 0px 1px #FF7079;
}

.c6::-webkit-input-placeholder {
  color: #A0A0A0;
  font-weight: 500;
}

.c6::-moz-placeholder {
  color: #A0A0A0;
  font-weight: 500;
}

.c6:-ms-input-placeholder {
  color: #A0A0A0;
  font-weight: 500;
}

.c6::placeholder {
  color: #A0A0A0;
  font-weight: 500;
}

.c4 h4 {
  margin-bottom: 10px;
  color: #A0A0A0;
}

.c5 {
  color: #A0A0A0;
}

.c8 {
  border-radius: 3px;
  font-weight: bold;
  font-size: 14px;
  outline: none;
  text-align: center;
  background: #FFFFFF;
  -webkit-transition: color ease .2s;
  transition: color ease .2s;
  -webkit-transition: background ease .2s;
  transition: background ease .2s;
  width: 100%;
  border: 2px solid #FF7079;
  color: #FF7079;
  cursor: pointer;
}

.c8:hover,
.c8:focus {
  background: #FF7079;
  color: #FFFFFF !important;
}

.c8:hover svg,
.c8:focus svg {
  fill: #FFFFFF;
}

.c8:active {
  background: #EF6069;
}

.c11 {
  color: #FF7079;
  font-size: 14px;
  padding: 2px;
  cursor: pointer;
}

.c11:hover {
  background: #FF7079;
  color: #FFFFFF;
}

.c10 {
  color: #FF7079;
}

.c1 {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  height: 23px;
  z-index: 3;
  -webkit-app-region: drag;
}

.c0 {
  display: grid;
  grid-template-rows: repeat(4,25vh);
  grid-template-columns: repeat(6,minmax(100px,1fr));
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
}

.c2 {
  padding: 40px;
  grid-column: 2 / 6;
  grid-row: 2 / 4;
  min-width: 300px;
}

.c3 {
  text-align: center;
  font-size: 40px;
  color: #FF7079;
}

.c9 {
  grid-row: 4;
  grid-column: 2 / 6;
  -webkit-align-self: end;
  -ms-flex-item-align: end;
  align-self: end;
  margin-bottom: 20px;
}

.c7 {
  padding: 10px 5px;
  margin: 25px auto 0px auto;
}

<div
  className="c0"
>
  <div
    className="c1"
  />
  <form
    className="c2"
    onSubmit={[Function]}
  >
    <h1
      className="c3"
    >
      Redshape
    </h1>
    <div
      className="form-group undefined c4"
    >
      <h4
        className="c5"
        htmlFor="username"
      >
        Login
      </h4>
      <input
        className="c6"
        disabled={false}
        name="username"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group undefined c4"
    >
      <h4
        className="c5"
        htmlFor="password"
      >
        Password
      </h4>
      <input
        className="c6"
        disabled={false}
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        type="password"
        value=""
      />
    </div>
    <div
      className="form-group undefined c4"
    >
      <h4
        className="c5"
        htmlFor="redmineEndpoint"
      >
        Redmine Endpoint
      </h4>
      <input
        className="c6"
        disabled={false}
        name="redmineEndpoint"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="https://redmine.example.com"
        type="text"
        value=""
      />
    </div>
    <button
      className="c7 c8"
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </form>
  <div
    className="c9"
  >
    <p
      className="c10"
    >
      Created byÂ 
      <a
        className="c11"
        href="https://spring3.github.io/website/"
        onClick={[Function]}
        rel="noopener noreferer"
        type="external"
      >
        Daniyil Vasylenko
      </a>
    </p>
  </div>
</div>
`;
